samples/drivers/display/src/main.c

```diff
@@ -0,0 +1,272 @@
+			       size_t buf_size)
+{
+	u16_t color = 0;
+	u16_t grey_5bit;
+
+	switch (corner) {
+	case TOP_LEFT:
+		color = 0xF800u;
+		break;
+	case TOP_RIGHT:
+		color = 0x07E0u;
+		break;
+	case BOTTOM_RIGHT:
+		color = 0x001Fu;
+		break;
+	case BOTTOM_LEFT:
+		grey_5bit = grey & 0x1Fu;
+		/* shift the green an extra bit, it has 6 bits */
+		color = grey_5bit << 11 | grey_5bit << (5 + 1) | grey_5bit;
+		break;
+	}
+
+	for (size_t idx = 0; idx < buf_size; idx += 2) {
+		*(buf + idx + 0) = (color >> 8) & 0xFFu;
+		*(buf + idx + 1) = (color >> 0) & 0xFFu;
+	}
+}
+
+static void fill_buffer_mono(enum corner corner, u8_t grey, u8_t *buf,
+			     size_t buf_size)
+{
+	u16_t color;
+
+	switch (corner) {
+	case BOTTOM_LEFT:
+		color = (grey & 0x01u) ? 0xFFu : 0x00u;
+		break;
+	default:
+		color = 0;
+		break;
+	}
+
+	memset(buf, color, buf_size);
+}
+
+void main(void)
+{
+	size_t x;
+	size_t y;
+	size_t rect_w;
+	size_t rect_h;
+	size_t h_step;
+	size_t grey_count;
+	u8_t *buf;
+	s32_t grey_scale_sleep;
+	struct device *display_dev;
+	struct display_capabilities capabilities;
+	struct display_buffer_descriptor buf_desc;
+	size_t buf_size = 0;
+	fill_buffer fill_buffer_fnc = NULL;
+
+	LOG_INF("Display sample for %s", DISPLAY_DEV_NAME);
+
+	display_dev = device_get_binding(DISPLAY_DEV_NAME);
+
+	if (display_dev == NULL) {
+		LOG_ERR("Device not found. Aborting sample.");
+		RETURN_FROM_MAIN(1);
+	}
+
+	display_get_capabilities(display_dev, &capabilities);
+
+	rect_w = capabilities.x_resolution / 4;
+	rect_h = capabilities.y_resolution / 8;
+
+	if ((capabilities.screen_info & SCREEN_INFO_MONO_VTILED) &&
+	    (rect_h % 8)) {
+		rect_h += 8 - (rect_h % 8);
+	}
+
+	if (capabilities.screen_info & SCREEN_INFO_EPD) {
+		grey_scale_sleep = K_MSEC(10000);
+	} else {
+		grey_scale_sleep = K_MSEC(100);
+	}
+
+	switch (capabilities.current_pixel_format) {
+	case PIXEL_FORMAT_ARGB_8888:
+		fill_buffer_fnc = fill_buffer_argb8888;
+		buf_size = 4 * rect_w * rect_h;
+		break;
+	case PIXEL_FORMAT_RGB_888:
+		fill_buffer_fnc = fill_buffer_rgb888;
+		buf_size = 3 * rect_w * rect_h;
+		break;
+	case PIXEL_FORMAT_RGB_565:
+		fill_buffer_fnc = fill_buffer_rgb565;
+		buf_size = 2 * rect_w * rect_h;
+		break;
+	case PIXEL_FORMAT_MONO01:
+	case PIXEL_FORMAT_MONO10:
+		fill_buffer_fnc = fill_buffer_mono;
+		buf_size = rect_w * rect_h / 8;
+		break;
+	default:
+		LOG_ERR("Unsupported pixel format. Aborting sample.");
+		RETURN_FROM_MAIN(1);
+	}
+
+	buf = k_malloc(buf_size);
+
+	if (buf == NULL) {
+		LOG_ERR("Could not allocate memory. Aborting sample.");
+		RETURN_FROM_MAIN(1);
+	}
+
+	(void)memset(buf, 0xFFu, buf_size);
+	h_step = (rect_w * rect_h) / capabilities.x_resolution;
+
+	buf_desc.buf_size = buf_size;
+	buf_desc.pitch = capabilities.x_resolution;
+	buf_desc.width = capabilities.x_resolution;
+	buf_desc.height = h_step;
+
+	for (int idx = 0; idx < capabilities.y_resolution; idx += h_step) {
+		display_write(display_dev, 0, idx, &buf_desc, buf);
```

Comment:
I have tested it with ssd1306, reel board, and adafruit_2_8_tft_touch_v2 for a while. Monochrome displays do not work with this sample if the h_step is not a multiple of an octet, there are also few issues in the drivers. I got it working as intended on ssd1306 with the changes below. I also suggest to define rect_w and rect_h as minimally addressable area for a display type and the scale it up. Also I observed that this sample did not work on ili9340 display on the Adafruit shield until I flashed it with samples/gui/lvgl first, maybe an initialization issue in the driver self.

```diff
diff --git a/samples/drivers/display/src/main.c b/samples/drivers/display/src/main.c
index b74c3c557a..ea7481e321 100644
--- a/samples/drivers/display/src/main.c
+++ b/samples/drivers/display/src/main.c
@@ -154,7 +154,7 @@ void main(void)
        size_t y;
        size_t rect_w;
        size_t rect_h;
-       size_t h_step;
+       size_t h_step = 0;
        size_t grey_count;
        u8_t *buf;
        s32_t grey_scale_sleep;
@@ -169,7 +169,8 @@ void main(void)
        display_dev = device_get_binding(DISPLAY_DEV_NAME);
 
        if (display_dev == NULL) {
-               LOG_ERR("Device not found. Aborting sample.");
+               LOG_ERR("Device %s not found. Aborting sample.",
+                       DISPLAY_DEV_NAME);
                RETURN_FROM_MAIN(1);
        }
 
@@ -193,19 +194,23 @@ void main(void)
        case PIXEL_FORMAT_ARGB_8888:
                fill_buffer_fnc = fill_buffer_argb8888;
                buf_size = 4 * rect_w * rect_h;
+               h_step = (rect_w * rect_h) / capabilities.x_resolution;
                break;
        case PIXEL_FORMAT_RGB_888:
                fill_buffer_fnc = fill_buffer_rgb888;
                buf_size = 3 * rect_w * rect_h;
+               h_step = (rect_w * rect_h) / capabilities.x_resolution;
                break;
        case PIXEL_FORMAT_RGB_565:
                fill_buffer_fnc = fill_buffer_rgb565;
                buf_size = 2 * rect_w * rect_h;
+               h_step = (rect_w * rect_h) / capabilities.x_resolution;
                break;
        case PIXEL_FORMAT_MONO01:
        case PIXEL_FORMAT_MONO10:
                fill_buffer_fnc = fill_buffer_mono;
-               buf_size = rect_w * rect_h / 8;
+               buf_size = capabilities.x_resolution * rect_h / 8;
+               h_step = rect_h;
                break;
        default:
                LOG_ERR("Unsupported pixel format. Aborting sample.");
@@ -220,7 +225,6 @@ void main(void)
        }
 
        (void)memset(buf, 0xFFu, buf_size);
-       h_step = (rect_w * rect_h) / capabilities.x_resolution;
 
        buf_desc.buf_size = buf_size;
        buf_desc.pitch = capabilities.x_resolution;
```

Reply:
@jfischer-phytec-iot thx for testing!

>  I got it working as intended on ssd1306 with the changes below.

Will take the changes along

>  I also suggest to define rect_w and rect_h as minimally addressable area for a display type and the scale it up. 

I will update the code

> Also I observed that this sample did not work on ili9340 display on the Adafruit shield until I flashed it with samples/gui/lvgl first, maybe an initialization issue in the driver self.

Hmm strange I also tested in a `adafruit_2_8_tft_touch_v2` shield and did not see any issue.
I will repeat the test and see what comes out of it.

Reply:
@jfischer-phytec-iot updated the code

On `adafruit_2_8_tft_touch_v2` shield the sample works for me out of the box* with the exception that the top left pixels of the rectangle are not correct

_*_ Tested with a `nrf52840_pca10056` preview board so I had to update the dts to use SPI 1 instead of 3